### PR TITLE
Add accessibility checking

### PR DESCRIPTION
**Pull request by pweaver-google**
Add accessibility checking, both as an explicit call and as an option in ShadowView.checkedPerformClick.

### CODE REVIEW

**Comment by jongerrish** [comment]
Any chance you can squash all those commits?

**Comment by jongerrish** [comment]
Did you think about adding the accessibility checks some place other than the onClick of a view(), I wonder if having them in the layout inflater, or when a view is attached to the window perhaps?

**Comment by erd** [comment]
Could you provide a little more context on what the change entails, since it looks like this is a new feature. Why would I want to do this in a test as opposed to catching this in the linter?

File: robolectric-annotations/src/main/java/org/robolectric/annotation/AccessibilityChecks.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+   *  option forces Robolectric to run only those checks that were present in a
+   *  specified version of Robolectric, which reduces the opportunity for a
+   *  new bug to be discovered in an old test. Note that this option does not
+   *  guarantee that the behavior of individual checks won't change as bugs are
+   *  fixed and/or features added to checks that existed in the specified 
+   *  version.
+   *  <p>
+   *  The default is to perform all available checks.
+   */
+  ForRobolectricVersion forRobolectricVersion() default ForRobolectricVersion.LATEST;
```
Comment (erd) [inline robolectric-annotations/src/main/java/org/robolectric/annotation/AccessibilityChecks.java]: Is this used anywhere? I see it declared in the annotation, but not used in any of the other files.

File: robolectric-utils/src/main/java/org/robolectric/util/AccessibilityUtil.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+   * 
+   * @param view The {@code View} to examine
+   * @return {@code false} if accessibility checks are enabled and a problem
+   * was found. {@code true} otherwise.
+   */
+  public static boolean passesAccessibilityChecksIfEnabled(View view) {
+    boolean checksEnabled = false;
+
+    /* Pull defaults from environment variables */
+    String checksEnabledString = System.getenv("a11y_checks_enabled");
```
Comment (erd) [inline robolectric-utils/src/main/java/org/robolectric/util/AccessibilityUtil.java]: This should be named something like "robolectric.accessibility.check" to match the other system properties that Robolectric uses. See http://robolectric.org/configuring/ for other configuration options.

**Comment by pweaver-google** [comment]
> Could you provide a little more context on what the change entails, since it looks like this is a new feature. Why would I want to do this in a test as opposed to catching this in the linter?

Linting can catch some, but not all, errors. For example, the linter won't complain about an ImageView with a contentDescription set to null, because setting to null indicates either that no description is needed (because the View is purely decorative) or that one will be added dynamically. By checking any View where a user click is simulated, we can catch Views inappropriately marked as decorative or cases where the developer has neglected to provide a dynamic content description before he expects users to interact with the View.

Another example is checking the contrast on a View, which can't be done with lint.

**Comment by pweaver-google** [comment]
> Did you think about adding the accessibility checks some place other than the onClick of a view(), I wonder if having them in the layout inflater, or when a view is attached to the window perhaps?

I may want to expose other ways to check Views in the future, but I want to be careful to avoid false positives. Until the View is in its final state, I don't want to check it. I'm making the assumption that when a View is clicked on, it is both something users should interact with and in its final state.

File: robolectric-annotations/src/main/java/org/robolectric/annotation/AccessibilityChecks.java (hunk 1 above)
Comment (pweaver-google) [inline robolectric-annotations/src/main/java/org/robolectric/annotation/AccessibilityChecks.java]: > Is this used anywhere? I see it declared in the annotation, but not used in any of the other files.

Removed. This is a feature I plan to add but pulled back after deciding I want to think a bit more about how to expose it.

**Comment by pweaver-google** [comment]
> Any chance you can squash all those commits?

Somehow everything from my merges is still showing up as a commit. I'm concerned that squashing others' commits will do something weird. If I need to squash them, I'd like some reassurance that I'm doing it correctly.

File: robolectric-utils/src/main/java/org/robolectric/util/AccessibilityUtil.java (hunk 2 above)
Comment (pweaver-google) [inline robolectric-utils/src/main/java/org/robolectric/util/AccessibilityUtil.java]: Done.

**Comment by erd** [comment]
I'm not sure exactly what happened with the history in this PR, so I just merged it manually.

Next time you submit a PR, start with a fresh tree (delete it and re-clone), and do your work on a branch. Here's a good tutorial on how to rebase your pull request onto new changes from master: https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request

**Pull request closed without merging**